### PR TITLE
Make the tool_lints actually usable

### PR DESCRIPTION
**Pull request by flip1995**
cc #44690

Necessary for rust-lang-nursery/rust-clippy#2955 and rust-lang-nursery/rust-clippy#2977

This PR makes it possible for lint tools (at the moment only for Clippy) to implement the `tool_lints`, like it was documented in #52018.

Because the `declare_lint` macro is pretty cluttered right now, there was not really a good way to add the `tool_name` as an additional argument of the macro. That's why I chose to introduce the new `declare_tool_lint` macro.

The switch from `&str` to `String` in the `lint_groups` `FxHashMap` is because I got weird error messages in the `check_lint_name` method. And the `by_name` field of the `LintStore` also uses `String`.

### What comes with this PR:

If this PR lands and Clippy switches to the `tool_lints`, the currently used methods
```rust
#[cfg_attr(feature = "cargo-clippy", allow(clippy_lint))]
#[allow(unknown_lints, clippy_lint)]
```
to `allow`/`warn`/`deny`/`forbid` Clippy lints, won't have any effects anymore, but also won't produce a warning. That is because the name of `clippy_lint` will then be `clippy::clippy_lint`. (Maybe we can add a clippy lint to search for `cfg_attr` appearances with the `cargo-clippy` feature?)

r? @oli-obk

### CODE REVIEW

File: src/librustc/lint/levels.rs [hunk 1]
```diff
@@ -247,17 +248,36 @@ impl<'a> LintLevelsBuilder<'a> {
...
+                        if let Some(ids) = result {
+                            let complete_name = &format!("{}::{}", tool_name.unwrap(), name);
+                            let src = LintSource::Node(Symbol::intern(complete_name), li.span);
+                            for id in ids {
+                                specs.insert(*id, (level, src));
+                            }
+                        }
+                        //FIXME: if Tool(None) is returned than the lint either does not exist in
+                        //the lint tool or the code doesn't get compiled with the lint tool and
+                        //therefore the lint cannot exist.
```
Comment (flip1995) [inline src/librustc/lint/levels.rs]: I'm not sure what to do here... Can we somehow detect if we currently use the `clippy_driver` or the `rustc_driver` to find out if the lint does not exists or wasn't even added in the first place?

Comment (oli-obk) [reply src/librustc/lint/levels.rs]: The driver could register itself as a tool. Not sure where that information belongs, maybe a new field in the `Session`?

Comment (oli-obk) [reply src/librustc/lint/levels.rs]: I think tool attributes are required to be checked by the tool. So maybe instead we leave the "unknown lint" error to clippy?

File: src/librustc/lint/context.rs [hunk 2]
```diff
@@ -69,7 +69,7 @@ pub struct LintStore {
 
     /// Map of registered lint groups to what lints they expand to. The bool
     /// is true if the lint group was added by a plugin.
-    lint_groups: FxHashMap<&'static str, (Vec<LintId>, bool)>,
+    lint_groups: FxHashMap<String, (Vec<LintId>, bool)>,
```
Comment (oli-obk) [inline src/librustc/lint/context.rs]: What's the motivation for this change?

Comment (flip1995) [reply src/librustc/lint/context.rs]: In the `check_lint_name` function I copied the check for `lint_groups` in the match of the `tool_lints`:
https://github.com/rust-lang/rust/pull/52851/files#diff-771d6215d1fb5e7f5503514f9895f7bdR341

This lead to the error 
```bash
error[E0277]: the trait bound `&str: std::borrow::Borrow<std::string::String>` is not satisfied
   --> librustc/lint/context.rs:341:48
    |
341 |                 None => match self.lint_groups.get(&complete_name) {
    |                                                ^^^ the trait `std::borrow::Borrow<std::string::String>` is not implemented for `&str`
```

The weird part was that this error appeared only once and not in both cases. But I can't reproduce this anymore. Switching to `String` for the `FxHashMap` keys solved the problem.

---

While thinking about this a little longer and not at midnight, I found an easier solution...

**Comment by flip1995** [comment]
I will squash the last 3 commits before merging.

File: src/librustc/lint/levels.rs (hunk 1 above)
Comment (flip1995) [reply src/librustc/lint/levels.rs]: Leaving it to the tool/Clippy was also my idea. We basically need a lint, which looks for `allow(clippy::_)`/... and checks if the lint exists, I think.

Comment (oli-obk) [reply src/librustc/lint/levels.rs]: So... I guess just turn the FIXME comment into a comment explaining that

**Review by oli-obk — APPROVED** [review]
r=me with comment adjusted and fixup commits squashed

File: src/librustc/lint/levels.rs (hunk 1 above)
Comment (flip1995) [reply src/librustc/lint/levels.rs]: Ok will do. And I will squash these commits while doing this.
